File: calculator2/Add.php
<?php

class Add implements Expression{

    private Expression $a;
    private Expression $b;

    public function __construct(Expression $a , Expression $b){
        $this->a = $a;
        $this->b = $b;
    }

    //式を評価した結果と式を評価した結果をreturnする
    public function evaluate(){
        return $this->a->evaluate() + $this->b->evaluate();
    }
}

File: calTeacherCode/Add.java

public class Add extends BinaryOperator {
	
	private Expression left;
	private Expression right;
	
	public Add(Expression a, Expression b) {
		this.left = a;
		this.right = b;
	}
	
	@Override
	public int evaluate() {
		if (left.isNull() || right.isNull()) {
			throw new RuntimeException();
		}
		return left.evaluate() + right.evaluate();
	}

	@Override
	public boolean isNull() {
		return false;
	}
	
	public static Expression empty() {
		return new Add(new NullValue(), new NullValue());
	}
	
	public void setLeft(Expression left) {
		this.left = left;
	}

	public void setRight(Expression right) {
		this.right = right;
	}
}
